gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,270 @@
+            w_x = track.point3()
+            for k1 in range(num_measurements):
+                i_a = measurements[k1][0]
+                # Check if i_a is a valid image with estimated camera pose, then get its id for patchmatchnet
+                i_pm_a = self._camera_idx_to_patchmatchnet_idx.get(i_a, -1)
+
+                # Calculate track_i's depth in the camera pose
+                z_a = self._sfm_result.get_camera(i_a).pose().transformTo(w_x)[-1]
+                # Update image i's depth list only when i in a valid camera
+                depths[i_pm_a].append(z_a)
+
+                for k2 in range(k1 + 1, num_measurements):
+                    i_b = measurements[k2][0]
+
+                    # Check if i_b is a valid image with estimated camera pose, then get its id for patchmatchnet
+                    i_pm_b = self._camera_idx_to_patchmatchnet_idx.get(i_b, -1)
+
+                    # Calculate the pairwise gaussian score only if i_a and i_b are both images with estimated poses
+                    if i_pm_a < 0 or i_pm_b < 0:
+                        continue
+
+                    # If both cameras are valid cameras,
+                    #   calculate score for track_i in the pair views (cam_a, cam_b)
+                    score_a_b = piecewise_gaussian(
+                        xPa=self._camera_centers[i_a] - w_x, xPb=self._camera_centers[i_b] - w_x
+                    )
+                    # Sum up pair scores for each track_i
+                    pair_scores[i_pm_b, i_pm_a] += score_a_b
+                    pair_scores[i_pm_a, i_pm_b] += score_a_b
+
+        # Sort pair scores, for i-th row, choose the largest (num_views-1) scores, the corresponding views are selected
+        #   as (num_views-1) source views for i-th reference view.
+        pairs = np.argsort(-pair_scores, axis=1)[:, : self._num_views - 1]
+
+        # Filter out depth outliers and calculate depth ranges
+        depth_ranges = np.zeros((self._num_images_cal, 2))
+        for i in range(self._num_images_cal):
+            depth_ranges[i, 0] = np.percentile(depths[i], MIN_DEPTH_PERCENTILE)
+            depth_ranges[i, 1] = np.percentile(depths[i], MAX_DEPTH_PERCENTILE)
+
+        return pairs, depth_ranges
+
+    def __len__(self) -> int:
+        """Get the length of the dataset
+        In the PatchMatchNet workflow, each input image will be treated as a reference view once, and each input image
+        corresponds to a estimated camera pose. So the length of the dataset is equal to the number of reference views.
+
+        Returns:
+            the length of the dataset
+        """
+        return self._num_images_cal
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        """Get inference data to PatchmatchNet
+        Produces data containing _num_views images, the first images is the reference image
+
+        Args:
+            index: index of yield test data, the reference image ID of the test data is _keys[index]
```

Comment:
can you please explain it to me a bit more. i did not understand the sentence.

Reply:
there're 2 sets of index systems, one is the camera id in SfMResult (might be discontinuous because of invalid cameras), and the other is the continuous view id (from 0 to the `number_of_valid_images - 1`) for PatchmatchNet. Here we hope to tell the user that the `index` here is the second one.

Reply:
Is this detailed anywhere in the code? If not you should add this somewhere.